### ISSUE TITLE
Beastling creates Beast XML with NullPointerException

### ISSUE THREAD

**Issue by Anaphory** [lmaurits/BEASTling#114]
Presumably due to a comma in a feature name, Beast does die on a Beastling-generated XML file, saying the following:

```
simulation-0-2/simulation-0-2.xml
Checking out /home/gereon/.beast/2.4/BEAST/lib
jardir = /home/gereon/beasties/beast2/build/dist/launcher.jar
Checking out /home/gereon/beasties/beast2/build/dist
Loaded URL file:/home/gereon/beasties/beast2/build/dist/beast.jar
java.lang.NullPointerException
	at beast.app.beastapp.BeastLauncher.copyFileUsingStream(Unknown Source)
	at beast.app.beastapp.BeastLauncher.createBeastPackage(Unknown Source)
	at beast.app.beastapp.BeastLauncher.loadBEASTJars(Unknown Source)
	at beast.app.beastapp.BeastLauncher.main(Unknown Source)

                   BEAST v2.4.4 Prerelease, 2002-2016
             Bayesian Evolutionary Analysis Sampling Trees
                       Designed and developed by
 Remco Bouckaert, Alexei J. Drummond, Andrew Rambaut & Marc A. Suchard
                                    
                     Department of Computer Science
                         University of Auckland
                        remco@cs.auckland.ac.nz
                        alexei@cs.auckland.ac.nz
                                    
                   Institute of Evolutionary Biology
                        University of Edinburgh
                           a.rambaut@ed.ac.uk
                                    
                    David Geffen School of Medicine
                 University of California, Los Angeles
                           msuchard@ucla.edu
                                    
                      Downloads, Help & Resources:
                           http://beast2.org/
                                    
  Source code distributed under the GNU Lesser General Public License:
                   http://github.com/CompEvol/beast2
                                    
                           BEAST developers:
   Alex Alekseyenko, Trevor Bedford, Erik Bloomquist, Joseph Heled, 
 Sebastian Hoehna, Denise Kuehnert, Philippe Lemey, Wai Lok Sibon Li, 
Gerton Lunter, Sidney Markowitz, Vladimir Minin, Michael Defoin Platel, 
                 Oliver Pybus, Chieh-Hsi Wu, Walter Xie
                                    
                               Thanks to:
          Roald Forsberg, Beth Shapiro and Korbinian Strimmer

Random number seed: 1481113017178

File: simulation-0-2.xml seed: 1481113017178 threads: 1

Error 170 parsing the xml input file

nullCould not find object associated with idref featureClockRate:meaningcognates:admit, confess

Error detected about here:
  <beast>
      <run id='mcmc' spec='MCMC'>
          <distribution id='posterior' spec='util.CompoundDistribution'>
              <distribution id='likelihood' spec='util.CompoundDistribution'>
                  <distribution id='featureLikelihood:meaningcognates:admit, confess' spec='TreeLikelihood'>
                      <siteModel id='SiteModel.meaningcognates:admit, confess' spec='SiteModel'>
                          <input>

evaluation/run_beast.sh: line 18: simulation-0-2/beast_simulation-0-2.tre: No such file or directory
```

Relevant files are attached. Beastling call was `cat vocabulary.tsv | beastling beastling.ini -o simulation-0-2.xml`.

**Comment by Anaphory** [comment]
Hm. I added some beastrun tests with commas in Feature IDs, but they succeed.

**Comment by Anaphory** [comment]
They should not succeed, because commas mess with plates in the beast XML and the above run on simulation-0-2.xml exits with non-zero status, so the beastrun *should* fail.

Pull request for those tests to follow, maybe someone else can make more sense of them.

**Comment by lmaurits** [comment]
In my local checkout, I've updated the new `validate_ids` method of `BeastXml` and now, at long last, these beastrun tests throw errors like they always should have!

But what now?  We _could_ simply move these configs into `bad_configs` so that they are expected to throw errors and close this issue.  But might it make more sense to add something like a `normalise_ids` method to `BeastXml` which does some simple transformations (replacing commas and whitespace with underscores, for example) so that these kind of IDs are guaranteed never to make it to an XML file and we don't have to test for / fix them in every possible location the might be generated...

**Comment by Anaphory** [comment]
Fixing errors sounds much better than tracking those things. I mean, it should not be part of our public API that certain configurations with commas in IDs should fail – undefined behaviour was fine until now that we have a nice infrastructure to fix them.

**Comment by xrotwang** [comment]
In #247 a added a correction for the "comma-in-feature name" issue, a small helper function `xml.valid_id` which is now used exclusively for `id` and `idref` in the `featureLikelihood` case.
Maybe it should be used for **all** `id` and `idref` attributes?

**Comment by xrotwang** [comment]
I just did that, i.e. run **all** `id`, `idref` and `range` attributes through `xml.valid_id` - and at least this doesn't break any tests. Should I add this behaviour to #247 ?

**Comment by lmaurits** [comment]
I think with the new running everything through `valid_id` this issue should be pretty much solved, so I'll close this (more than 2.5 years later!!!).
